smolyay/samples.py

```diff
@@ -0,0 +1,496 @@
+    .. math::
+
+        n = \begin{cases}
+            0 & \text{ if } L = 0\\ 
+            2^{L} & \text{ if } L > 0 
+        \end{cases}
+
+    where L is a whole number.
+
+    As the extrema of the Chebyshev polynomials of degree :math:`n` are nested,
+    meaning any :math:`n` will have extrema at the same points as every n that
+    precedes it in the sequence. The extrema are organized into levels such that
+    a level L will contain the extrema of Chebyshev polynomial n(L) that are
+    not the extrema of any Chebyshev polynomial n(k) where k is a whole number
+    and k < L.
+
+    The order :math:`o(L) of the nested Clenshaw Curtis set, which describes the
+    number of cummulative points at each level, is used to determine the 
+    number of points at each individual level. The order of the Clenshaw Curtis
+    set is described by the following rule:
+
+    .. math::
+
+        o(L) = \begin{cases}
+                1 & \text{ if } L = 0\\ 
+                2^{L} + 1 & \text{ if } L > 0 
+        \end{cases}
+
+    which leads to a sequence :math:`{1, 3, 5, 9, 17, ...}`.
+
+    Determining the number of points each level is then
+
+    .. math::
+        num_per_level(L) = o(L) - o(L - 1)
+
+    Parameters
+    ----------
+    domain: list
+        Domain of the sample points.
+
+    max_level ; int
+        The maximum level the points are used for.
+    """
+
+    def _create(self):
+        r"""Create the points in the set.
+
+        Generating nested extrema of chebyshev polynomials of the first kind.
+        """
+        # create properties for levels, level 0 is a special case with 1 point
+        rule = lambda x: 1 if x == 0 else 2**x + 1
+        self._num_per_level = numpy.ones(self.max_level + 1, dtype=int)
+        self._num_per_level[1:] = [
+            rule(i) - rule((i - 1)) for i in range(1, self.max_level + 1)
+        ]
+        self._end_level = numpy.cumsum(self._num_per_level)
+        self._start_level = self._end_level - self._num_per_level
+
+        # points, level 0 is a special case only 0 as a point
+        num_points = self._end_level[-1]
+        points = numpy.zeros(num_points, dtype=float)
+        for i in range(1, self.max_level + 1):
+            # find indexes of extrema not already found. Fraction index/degree
+            # cannot be further simplified
+            degree = 2**i
+            if i == 1:
+                # special case for level == 1
+                indexes = numpy.arange(0, degree + 1, 2, dtype=int)
+            else:
+                indexes = numpy.arange(1, degree, 2, dtype=int)
+                divisible_indexes = numpy.gcd(indexes, degree) > 1
+                indexes = indexes[~divisible_indexes]
+            points[self._start_level[i] : self._end_level[i]] = -numpy.cos(
+                numpy.pi * indexes / degree
+            )
+        self._points = self._scale_to_domain(points, [-1, 1])
+
+
+class SlowNestedClenshawCurtisPointSet(NestedClenshawCurtisPointSet):
```

Comment:
Should this inherit from NestedClenshawCurtisPointSet or NestedUnidimensionalPointSet? I would say the second one, since a SlowNestedClenshawCurtisPointSet is not really a NestedClenshawCurtisPointSet in the way we have defined them, and it doesn't actually use anything from that class either.

Reply:
It originally inherited from the NestedClenshawCurtisPointSet under the assumption that calculating certain things like points would be separate methods that the slow version could inherit, It should inherit from the NestedUnidimensionalPointSet since that's not the case anymore.